### PR TITLE
Add Communicator.ActivateAsync

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR splits communicator creation in two phases:
- construction, which is naturally synchronous
- asynchronous activation, by calling ActivateAsync on the communicator object

This PR also renames IPlugin's InitializeAsync to ActivateAsync for consistency.

With the PR, the following configuration properties are no longer used:
- Ice.InitPlugins (since plug-in activation is now always explicit with ActivateAsync)
- Ice.Admin.DelayCreation (since Admin creation is always delayed until ActivateAsync)

### CODE REVIEW

File: csharp/src/Ice/Communicator-PluginManager.cs [hunk 1]
```diff
@@ -178,7 +163,7 @@ private void LoadPlugins(ref string[] cmdArgs)
                 }
             }
 
-            // Load and initialize the plug-ins defined in the property set with the prefix "Ice.Plugin.". These
+            // Loads and activates the plug-ins defined in the property set with the prefix "Ice.Plugin.". These
```
Comment (pepone) [inline csharp/src/Ice/Communicator-PluginManager.cs:166]: Fix the comment, we no longer initialize/activate the plugins here

File: csharp/src/Ice/Communicator.cs [hunk 2]
```diff
@@ -759,28 +757,27 @@ internal CancellationToken CancellationToken
...
         {
-            // An application can set Ice.InitPlugins=0 if it wants to postpone initialization until after it has
-            // interacted directly with the plug-ins.
-            if (GetPropertyAsBool("Ice.InitPlugins") ?? true)
+            if (_activateCalled)
             {
-                await InitializePluginsAsync(cancel).ConfigureAwait(false);
+                throw new InvalidOperationException("ActivateAsync was already called on this communicator");
             }
+            _activateCalled = true;
```
Comment (pepone) [inline csharp/src/Ice/Communicator.cs:771]: Why don't use the lock to set _activateCalled, this doesn't seem thread safe

**Review by pepone — APPROVED** [review]
Looks good, we should add a comment in config/PropertyNames.xml to remember removing properties that are no longer used

File: csharp/src/Ice/Communicator.cs (hunk 2 above)
Comment (bernardnormier) [reply csharp/src/Ice/Communicator.cs:771]: You're right, this is incorrect. Fixed.

File: csharp/src/Ice/Communicator-PluginManager.cs (hunk 1 above)
Comment (bernardnormier) [reply csharp/src/Ice/Communicator-PluginManager.cs:166]: Fixed.

File: csharp/src/Ice/Communicator-PluginManager.cs [hunk 3]
```diff
@@ -26,15 +25,15 @@ public sealed partial class Communicator
         private static readonly List<string> _loadOnInitialization = new List<string>();
         private static readonly Dictionary<string, IPluginFactory> _pluginFactories =
             new Dictionary<string, IPluginFactory>();
-        private bool _pluginsInitialized;
+
         private readonly List<(string Name, IPlugin Plugin)> _plugins = new List<(string Name, IPlugin Plugin)>();
```
Comment (bentoi) [inline csharp/src/Ice/Communicator-PluginManager.cs:29]: We chould use `new()` here for the init of the data members.

**Review by bentoi — APPROVED** [review]
Looks good. 

But now that we have a new "activated" state for the communicator, should we consider removing the `PluginActivationContext` class? 

It's not really new but this new explicit state also makes more prominent the question of what is allowed and what isn't before the communicator activation. Can I make a proxy invocation on a non-activated communicator? You can but with some limitations... internal locators aren't activated, no interceptors setup by plugins, the default logger is used. 

For interceptors it's actually a bit problematic because activating the communicator and concurrently invoking on a proxy isn't thread-safe.

Also, should we consider removing the RegisterPluginFactory static method which is no longer used and which is not that useful in C#?

**Comment by pepone** [comment]
>Also, should we consider removing the RegisterPluginFactory static method which is no longer used and which is not that useful in C#?

I added this for Xamarin to ensure the plug-ins were linked in iOS builds, we will not be able to support Xamarin with .NET 5, we can remove it and see if we need something like this when .NET 6 Xamarin support comes around.

File: csharp/src/Ice/Communicator-PluginManager.cs (hunk 3 above)
Comment (bernardnormier) [reply csharp/src/Ice/Communicator-PluginManager.cs:29]: Fixed.

**Comment by bernardnormier** [comment]
I will fix the other comments in a follow-up PR.

**Pull request merged**